[WebGPU Swift] https://webgpu.github.io/webgpu-samples/?sample=helloTriangle crashes on launch

mwyrzykowski opened the pull request:
#### 5f940dafb1bdab24d0c3104794b9df352d7070b3
<pre>
[WebGPU Swift] <a href="https://webgpu.github.io/webgpu-samples/?sample=helloTriangle">https://webgpu.github.io/webgpu-samples/?sample=helloTriangle</a> crashes on launch
<a href="https://bugs.webkit.org/show_bug.cgi?id=298318">https://bugs.webkit.org/show_bug.cgi?id=298318</a>
<a href="https://rdar.apple.com/159757436">rdar://159757436</a>

Reviewed by Tadeu Zagallo.

WebGPU Swift was missing some changes resulting in assertions or
other crashes.

* Source/WebCore/Modules/WebGPU/Implementation/WebGPUCommandEncoderImpl.cpp:
(WebCore::WebGPU::CommandEncoderImpl::beginRenderPass):
* Source/WebGPU/WebGPU/CommandEncoder.swift:
(WebGPU.timestampWriteIndex(_:defaultValue:offset:)):
(WebGPU.beginRenderPass(_:)):
(WebGPU.beginComputePass(_:)):
* Source/WebGPU/WebGPU/Device.h:
* Source/WebGPU/WebGPU/Device.mm:
(WebGPU::Device::Device):
(WebGPU::Device::getXRViewSubImageDepthTexture const):
* Source/WebGPU/WebGPU/QuerySet.h:
(WebGPU::QuerySet::counterSampleBuffer const): Deleted.
* Source/WebGPU/WebGPU/TextureView.h:
* Source/WebGPU/WebGPU/TextureView.mm:
(WebGPU::TextureView::rasterizationMapForSlice):
* Source/WebGPU/WebGPU/WebGPU.h:
(wgpuGetRenderPassDescriptorNextChainedStruct): Deleted.

Canonical link: <a href="https://commits.webkit.org/299527@main">https://commits.webkit.org/299527@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c120677661d3d42d3d9317218ffe26abdb715d63

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/119289 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/38976 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/29631 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/125527 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/71359 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/afbe142f-5ddc-4826-987b-e3be908aeefe) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/121166 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/39673 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/47557 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/90637 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/59975 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/764736dc-c70c-4456-8e09-510bdf73f2a5) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/122241 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/31646 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/106950 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/71056 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/67dc28d1-3731-4796-9a30-409b73a8af89) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/30689 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/25059 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/69176 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/101098 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/25250 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/128534 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/46207 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/34953 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/99204 "Found 1 new test failure: imported/w3c/web-platform-tests/service-workers/service-worker/redirected-response.https.html (failure)") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/46572 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/103149 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/98980 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25160 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/44443 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/22455 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/42774 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/46070 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/51770 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/45535 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/48885 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/47222 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski commented on Source/WebGPU/WebGPU/WebGPU.h:
```diff
@@ -1383,7 +1375,7 @@ typedef struct WGPUDeviceDescriptor {
 } WGPUDeviceDescriptor WGPU_STRUCTURE_ATTRIBUTE;
 
 typedef struct WGPURenderPassDescriptor {
-    WGPUChainedStruct const * nextInChain;
```
Instead of dealing with `nextInChain` and casting which is not needed, I just inlined the member

mwyrzykowski commented on Source/WebGPU/WebGPU/Device.mm:
```diff
@@ -309,17 +309,15 @@ static uint32_t computeAppleGPUFamily(id<MTLDevice> device)
 {
 #if ENABLE(WEBGPU_SWIFT)
     NSError *error = nil;
-    static std::once_flag onceFlag;
```
`m_nopVertexFunction` is a member variable, so `call_once` is not going to work since each new call to `requestDevice` creates a new Device.

Alternative is to make `m_nopVertexFunction` a static variable

tadeuzagallo replied:
Nice, I was just fixing this too

tadeuzagallo commented on Source/WebGPU/WebGPU/CommandEncoder.swift:
```diff
@@ -2089,15 +2124,17 @@ extension WebGPU.CommandEncoder {
             computePassDescriptor.sampleBufferAttachments[0].sampleBuffer = counterSampleBuffer != nil ? computePassDescriptor.sampleBufferAttachments[0].sampleBuffer : m_device.ptr().timestampsBuffer(m_commandBuffer, 2)
-            computePassDescriptor.sampleBufferAttachments[0].startOfEncoderSampleIndex = timestampWrites != nil ? Int(timestampWrites!.beginningOfPassWriteIndex) : 0
-            computePassDescriptor.sampleBufferAttachments[0].endOfEncoderSampleIndex = timestampWrites != nil ? Int(timestampWrites!.endOfPassWriteIndex) : 1
+            // computePassDescriptor.sampleBufferAttachments[0].startOfEncoderSampleIndex = Int(timestampWriteIndex(writeIndex: timestampWrites!.beginningOfPassWriteIndex, defaultValue: , offset: counterSampleBuffer.second))
```
Is this supposed to be commented out?

mwyrzykowski replied:
No, thank for noticing I will fix that

mwyrzykowski committed:
[WebGPU Swift] https://webgpu.github.io/webgpu-samples/?sample=helloTriangle crashes on launch
https://bugs.webkit.org/show_bug.cgi?id=298318
rdar://159757436

Reviewed by Tadeu Zagallo.

WebGPU Swift was missing some changes resulting in assertions or
other crashes.

* Source/WebCore/Modules/WebGPU/Implementation/WebGPUCommandEncoderImpl.cpp:
(WebCore::WebGPU::CommandEncoderImpl::beginRenderPass):
* Source/WebGPU/WebGPU/CommandEncoder.swift:
(WebGPU.timestampWriteIndex(_:defaultValue:offset:)):
(WebGPU.beginRenderPass(_:)):
(WebGPU.beginComputePass(_:)):
* Source/WebGPU/WebGPU/Device.h:
* Source/WebGPU/WebGPU/Device.mm:
(WebGPU::Device::Device):
(WebGPU::Device::getXRViewSubImageDepthTexture const):
* Source/WebGPU/WebGPU/QuerySet.h:
(WebGPU::QuerySet::counterSampleBuffer const): Deleted.
* Source/WebGPU/WebGPU/TextureView.h:
* Source/WebGPU/WebGPU/TextureView.mm:
(WebGPU::TextureView::rasterizationMapForSlice):
* Source/WebGPU/WebGPU/WebGPU.h:
(wgpuGetRenderPassDescriptorNextChainedStruct): Deleted.

Canonical link: https://commits.webkit.org/299527@main
